### ISSUE TITLE
Custom discrete probability in the multidimensional case

### ISSUE THREAD

**Issue by Princeofthebow** [scipy/scipy#13014]
Hi, I'm trying to understand how to define a custom rv_discrete for multivariate probabilities.

I found this in the documentation for a monodimensional pmf:

```
Examples

Custom made discrete distribution:


from scipy import stats
xk = np.arange(7)
pk = (0.1, 0.2, 0.3, 0.1, 0.1, 0.0, 0.2)
custm = stats.rv_discrete(name='custm', values=(xk, pk))

import matplotlib.pyplot as plt
fig, ax = plt.subplots(1, 1)
ax.plot(xk, custm.pmf(xk), 'ro', ms=12, mec='r')
ax.vlines(xk, 0, custm.pmf(xk), colors='r', lw=4)
plt.show() 
```
but it does not seem to work, nor I'm sure the code is correct as I'm new to Python, for multimensional probabilities:

```
p = [[1/6, 2/6, 0], [1/6, 1/6, 1/6]]
xks=[[1,2,3],[1,2,3]];
joint = stats.rv_discrete(name='Joint', values=(xks, a));
print(joint.pmf(xks))
```

How can I handle this?

**Comment by rlucas7** [comment]
AFAIK the rv_discrete is a *univariate* discrete distribution. 

https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.rv_discrete.html

The coverage of multivariate distributions has the usual ones like multivariate normal, multivariate t (next release) and a couple others, there is not a generic multivariate discrete distribution in 

https://github.com/scipy/scipy/blob/master/scipy/stats/_multivariate.py

In terms of helping us decide if to support that could you describe what you’d like to do with a generic multivariate discrete? 

-Lucas Roberts

**Comment by Princeofthebow** [comment]
Hi Lucas,             
              thank you for this.

The main use I would see for a generic multivariate discrete distribution is to be able to:

1. Define it from probabilities that I can supply. For example in the use case that triggered this request I have a discrete probability distribution in a data frame obtained from numerical simulations
2. The main benefit under my point of view would be to compute expectations and moments of this user defined distribution either along one of the dimensions or for the whole distribution. CDF, pdf etc are also useful functions like the rest available ins rv_discrete()

In other words, being able to define such a Rv would be beneficial for manipulating it in various ways.
I'm not sure if https://github.com/scipy/scipy/blob/master/scipy/stats/_multivariate.py might serve my needs - I'll take a look soon.

Let me know what you think and or if you know of other ways to do similar things.

**Comment by rlucas7** [comment]
> Let me know what you think and or if you know of other ways to do similar things(e.g. directly in Pandas?).

I can't think of anything-off the top of my head-directly in Pandas but maybe ask on stack overflow and tag if you prefer it directly in Pandas. 

For SciPy, the closest thing I can think of that we have at the moment would be kde:
https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.gaussian_kde.html
but that isn't discrete.

We do have a `binned_statistic_dd` which can be used as a `d` dimensional histogram if your statistic is a count. 
Assuming you divide all counts by the total number of counts you'd have a discrete distribution. 
We don't have anything that works w/that to generate expectations, cdf, pdf, though and this seems to be what you're use case needs.  

You can see a list of the currently supported multivariate distributions here: 
https://github.com/scipy/scipy/blob/master/scipy/stats/_multivariate.py#L18-L28
none of them appear to have what you're looking for implemented.

what you're describing sounds like a fair bit of work to implement and support so it might be better in the short term to implement something that directly meets you specific needs and report back findings here. We could then use those learnings to inform a potential implementation if the other maintainers agree/think it is in scope.
